### PR TITLE
Support general reductions in ParallelAccelerator

### PR DESCRIPTION
**Pull request by ehsantn**
This PR provides infrastructure for general reduction support in ParallelAccelerator. A variable that is updated by a function in the body of a parfor, using itself as one argument, is detected and supported as a reduction variable. The variable should hold the initial (neutral) reduction value before entering the loop, except for reduction functions that are known to ParallelAccelerator (e,g, `min`).

`functools.reduce()` operator is also supported but requires the initial value argument.

The following numpy functions are now parallelized: `min`, `max`, `argmin`, `argmax`.

### CODE REVIEW

File: numba/parfor.py [hunk 1]
```diff
@@ -1915,3 +2110,11 @@ def parfor_typeinfer(parfor, typeinferer):
 
 
 typeinfer.typeinfer_extensions[Parfor] = parfor_typeinfer
+
+@infer_global(reduce)
+class ReduceInfer(AbstractTemplate):
+    def generic(self, args, kws):
+        assert not kws
+        assert len(args) == 3
+        assert isinstance(args[1], types.Array)
```
Comment (stuartarchibald) [inline numba/parfor.py]: This restriction should be noted in the documentation, along with requiring an initial value.

File: numba/targets/arrayobj.py [hunk 2]
```diff
@@ -1975,6 +1975,10 @@ def dtype_type(context, builder, dtypety, dtypeval):
     # Just return a dummy opaque value
     return context.get_dummy_value()
 
+@lower_getattr(types.DType, 'kind')
```
Comment (stuartarchibald) [inline numba/targets/arrayobj.py]: Is this used?

File: numba/tests/test_parfors.py [hunk 3]
```diff
@@ -497,6 +497,58 @@ def test_impl(x, is_positive, N):
...
+
+        A = np.random.ranf(4)
+        self.check(test_impl, A)
+        A = np.random.randint(10, size=4).astype(np.int32)
+        self.check(test_impl, A)
+
+    @skip_unsupported
+    def test_max(self):
+        def test_impl(A):
+            return A.max()
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: I assume the np versions before canonicalisation has taken place also work (`np.max(A)`)? Perhaps they should also be tested?

File: numba/tests/test_parfors.py [hunk 4]
```diff
@@ -497,6 +497,58 @@ def test_impl(x, is_positive, N):
...
+        self.check(test_impl, A)
+        A = np.random.randint(10, size=4).astype(np.int32)
+        self.check(test_impl, A)
+
+    @skip_unsupported
+    def test_argmin(self):
+        def test_impl(A):
+            return A.argmin()
+
+        A = np.random.ranf(4)
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: From a numerical perspective, it would be good to test that value aliases work as expected. e.g. `A = np.array([1., 0., 2, 0., 3.])` should return `1` not `3`.

File: numba/tests/test_parfors.py [hunk 5]
```diff
@@ -784,6 +836,18 @@ def test_impl(n):
         self.prange_tester(test_impl, np.int32(4))
 
     @skip_unsupported
+    def test_prange14(self):
+        def test_impl(A):
+            s = 3
+            for i in range(len(A)):
+                s += A[i]*2
+            return s
+        # this tests reduction detection well since accumulated variable is
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: s/accumulated/the accumulated/

File: numba/tests/test_parfors.py [hunk 6]
```diff
@@ -784,6 +836,18 @@ def test_impl(n):
...
+        # initialized before the parfor and value accessed from array is updated
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: s/array/the array/

File: numba/types/__init__.py [hunk 7]
```diff
@@ -23,6 +23,8 @@
 Any = Phantom('any')
 undefined = Undefined('undefined')
 string = Opaque('str')
+code_type = Opaque('code')
+pyfunc_type = Opaque('pyfunc')
```
Comment (stuartarchibald) [inline numba/types/__init__.py]: Does this need to be `@typeof_impl.register`'d somewhere?

File: numba/parfor.py [hunk 8]
```diff
@@ -234,6 +267,32 @@ def run(self):
             get_parfor_params(self.func_ir.blocks)
         return
 
+    def _replace_parallel_functions(self, blocks):
```
Comment (stuartarchibald) [inline numba/parfor.py]: Docstring/comment?

File: numba/parfor.py [hunk 9]
```diff
@@ -260,6 +319,23 @@ def _convert_numpy(self, blocks):
                 new_body.append(instr)
             block.body = new_body
 
+    def _convert_reduce(self, blocks):
```
Comment (stuartarchibald) [inline numba/parfor.py]: Docstring/comment?

File: numba/parfor.py [hunk 10]
```diff
@@ -356,6 +432,15 @@ def _is_prange(self, func_var, call_table):
         call = call_table[func_var]
         return len(call) > 0 and (call[0] == 'prange' or call[0] == prange)
 
+    def _is_reduce_assign(self, inst):
```
Comment (stuartarchibald) [inline numba/parfor.py]: Docstring/comment?

File: numba/parfor.py [hunk 11]
```diff
@@ -378,6 +463,16 @@ def _make_index_var(self, scope, index_vars, body_block):
             raise NotImplementedError(
                 "Parfor does not handle arrays of dimension 0")
 
+    def _mk_parfor_loops(self, size_vars, scope, loc):
```
Comment (stuartarchibald) [inline numba/parfor.py]: Docstring/comment?

File: numba/parfor.py [hunk 12]
```diff
@@ -635,6 +719,18 @@ def _reduction_to_parfor(self, equiv_set, lhs, expr):
         call_name, mod_name = find_callname(self.func_ir, expr)
         args = expr.args
         kws = dict(expr.kws)
+        if call_name is 'min':
```
Comment (stuartarchibald) [inline numba/parfor.py]: `s/is/==/` test equality not identity?

File: numba/parfor.py [hunk 13]
```diff
@@ -635,6 +719,18 @@ def _reduction_to_parfor(self, equiv_set, lhs, expr):
...
+            arr = expr.args[0]
+            init_val = get_type_max_value(self.typemap[arr.name].dtype)
+            expr.args.insert(0, lambda a,b: min(a, b))
+            expr.args.append(ir.Const(init_val, arr.loc))
+            return self._reduce_to_parfor(equiv_set, lhs, expr)
+        if call_name is 'max':
```
Comment (stuartarchibald) [inline numba/parfor.py]: `s/is/==/` test equality not identity?

File: numba/ir_utils.py [hunk 14]
```diff
@@ -1234,3 +1239,256 @@ def find_const(func_ir, var):
     var_def = get_definition(func_ir, var)
     require(isinstance(var_def, ir.Const))
     return var_def.value
+
+def get_type_max_value(typ):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/ir_utils.py]: Sounds interesting. Sure, I'm interested in seeing how it works.

File: numba/ir_utils.py [hunk 15]
```diff
@@ -1234,3 +1239,256 @@ def find_const(func_ir, var):
...
+                block.body.append(ir.Jump(return_label, stmt.loc))
+                # remove cast of the returned value
+                for cast in casts:
+                    if cast.target.name == stmt.value.name:
+                        cast.value = cast.value.value
+            elif isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Expr) and stmt.value.op == 'cast':
+                casts.append(stmt)
+
+
+class IndexValue(object):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/ir_utils.py]: Sure, I'll move them.

File: numba/ir_utils.py [hunk 16]
```diff
@@ -1234,3 +1239,256 @@ def find_const(func_ir, var):
...
+        elif bw == 64:
+            lty = lir.DoubleType()
+        else:
+            raise NotImplementedError("llvmlite only supports 32 and 64 bit floats")
+        npty = getattr(numpy, 'float{}'.format(bw))
+        res = lir.Constant(lty, numpy.finfo(npty).max)
+    return impl_ret_untracked(context, builder, lty, res)
+
+def compile_to_numba_ir(mk_func, glbls, typingctx=None, arg_typs=None,
+                        typemap=None, calltypes=None):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/ir_utils.py]: Done.

File: numba/ir_utils.py [hunk 17]
```diff
@@ -1234,3 +1239,256 @@ def find_const(func_ir, var):
...
+            f_typemap.pop(a)
+        typemap.update(f_typemap)
+        calltypes.update(f_calltypes)
+    return f_ir
+
+def get_ir_of_code(glbls, fcode):
+    """
+    Compile a code object to get its IR.
+    """
+    #glbls = self.func_ir.func_id.func.__globals__
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/ir_utils.py]: Done.

File: numba/targets/arrayobj.py (hunk 2 above)
Comment (ehsantn) [reply numba/targets/arrayobj.py]: This was used in a previous implementation to find out if a type is float or integer (and get its min/max). I left this here since it doesn't hurt :)

File: numba/parfor.py [hunk 18]
```diff
@@ -1200,40 +1332,103 @@ def get_parfor_outputs(parfor, parfor_params):
...
+                break;
+    assert reduce_nodes, "Invalid reduction format"
+    return reduce_nodes
+
+def get_expr_args(expr):
+    if expr.op in ['binop', 'inplace_binop']:
+        return [expr.lhs.name, expr.rhs.name]
+    if expr.op == 'call':
+        return [v.name for v in expr.args]
+    raise NotImplementedError("expr args")
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py [hunk 19]
```diff
@@ -1200,40 +1332,103 @@ def get_parfor_outputs(parfor, parfor_params):
...
+    require(nodes[-2].target.name == nodes[-1].value.name)
+    acc_expr = nodes[-2].value
+    require(isinstance(acc_expr, ir.Expr) and acc_expr.op=='inplace_binop')
+    if acc_expr.fn == '+=':
+        return 0
+    if acc_expr.fn == '*=':
+        return 1
+    return None
+
+def get_reduce_nodes(name, nodes):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py [hunk 20]
```diff
@@ -1200,40 +1332,103 @@ def get_parfor_outputs(parfor, parfor_params):
...
+    for param, used_vars in param_uses.items():
+        # a parameter is a reduction variable if its value is used to update it
+        # check names since recursive parfors might have processed param already
+        if param in used_vars and param not in names:
+            names.append(param)
+            param_nodes[param].reverse()
+            reduce_nodes = get_reduce_nodes(param, param_nodes[param])
+            init_val = guard(get_reduction_init, reduce_nodes)
+            reductions[param] = (init_val, reduce_nodes)
+    #print("reduce vars:", names, reductions)
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py [hunk 21]
```diff
@@ -717,6 +802,53 @@ def _reduction_to_parfor(self, equiv_set, lhs, expr):
         # return error if we couldn't handle it (avoid rewrite infinite loop)
         raise NotImplementedError("parfor translation failed for ", expr)
 
+    def _reduce_to_parfor(self, equiv_set, lhs, expr):
+        #return ir.Assign(expr.args[2], lhs, lhs.loc)
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py [hunk 22]
```diff
@@ -717,6 +802,53 @@ def _reduction_to_parfor(self, equiv_set, lhs, expr):
         # return error if we couldn't handle it (avoid rewrite infinite loop)
         raise NotImplementedError("parfor translation failed for ", expr)
 
+    def _reduce_to_parfor(self, equiv_set, lhs, expr):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/parfor.py]: Done. Yes, this is the format for `reduce()` function of `functools`.

File: numba/parfor.py (hunk 13 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py (hunk 12 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py (hunk 11 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py (hunk 10 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/parfor.py (hunk 8 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/typing/arraydecl.py [hunk 23]
```diff
@@ -453,6 +453,14 @@ def resolve_type(self, ary):
         # Wrap the numeric type in NumberClass
         return types.NumberClass(ary.dtype)
 
+    def resolve_kind(self, ary):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/typing/arraydecl.py]: Not anymore but doesn't hurt to have it...

File: numba/types/__init__.py (hunk 7 above)
Comment (ehsantn) [reply numba/types/__init__.py]: `pyfunc_type` doesn't need registering since it is just used for output of `make_function` nodes in type inference. We don't have a use case other than this but we can add it later if necessary.

File: numba/tests/test_parfors.py [hunk 24]
```diff
@@ -497,6 +497,58 @@ def test_impl(x, is_positive, N):
         is_positive = np.zeros(N)
         self.check(test_impl, x, is_positive, N)
 
+    @skip_unsupported
+    def test_reduce(self):
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: The problem with direct lowering of `reduce` is that function arguments are not supported. We can transform it to a regular loop at some point in the pipeline though.

File: numba/tests/test_parfors.py (hunk 3 above)
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Yes, they work. I added the tests.

File: numba/tests/test_parfors.py (hunk 5 above)
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Done.

File: numba/tests/test_parfors.py (hunk 6 above)
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Done.

File: numba/tests/test_parfors.py (hunk 4 above)
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Thanks for the suggestion. Done.

File: numba/tests/test_parfors.py [hunk 25]
```diff
@@ -497,6 +497,58 @@ def test_impl(x, is_positive, N):
...
+        A = np.random.randint(10, size=4).astype(np.int32)
+        cpfunc = self.compile_parallel(test_impl, (numba.typeof(A),))
+        self.assertEqual(cpfunc.entry_point(A), test_impl(A))
+
+    @skip_unsupported
+    def test_min(self):
+        def test_impl(A):
+            return A.min()
+
+        A = np.random.ranf(4)
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Increased the size of input array to 211 (prime number to test corner cases).

[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: I changed the default index in `parallel` mode to 0 but we can explore the raise option as well.

[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Good catch! The `numba` global needs to be in the global variables of the inlined function for compilation. I just pushed a fix.

File: numba/targets/numbers.py [hunk 26]
```diff
@@ -494,6 +494,23 @@ def int_sign_impl(context, builder, sig, args):
 lower_builtin('>', types.boolean, types.boolean)(int_ugt_impl)
 lower_builtin('>=', types.boolean, types.boolean)(int_uge_impl)
 
+@lower_builtin('==', types.Const, types.Const)
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/targets/numbers.py]: This is intended to cover the comparison of string constants. I added some tests in `test_operators`.

File: numba/inline_closurecall.py [hunk 27]
```diff
@@ -93,147 +94,120 @@ def run(self):
...
-                args = args + [ir.Const(value=v, loc=loc)
-                               for v in defaults.value]
-            else:
-                raise NotImplementedError(
-                    "Unsupported defaults to make_function: {}".format(defaults))
-        _replace_args_with(callee_blocks, args)
-        debug_print("After arguments rename: ")
+def inline_closure_call(func_ir, glbls, block, i, callee, typingctx=None,
+                        arg_typs=None, typemap=None, calltypes=None):
+    """Inline the body of `callee` at its callsite (`i`-th instruction of `block`)
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/inline_closurecall.py]: Done.

File: numba/parfor.py (hunk 1 above)
Comment (ehsantn) [reply numba/parfor.py]: Done.

File: numba/ir_utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/ir_utils.py]: I moved them to `typing/builtins.py` and `targets/builtins.py`but not sure if they fully fit.

File: numba/tests/test_parfors.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: I added supported for `reduce()` in njit using the inlining pass.

**Comment by ehsantn** [comment]
@stuartarchibald Thanks for the suggestions. I think I addressed all of them. Let me know what you think.

**Comment by stuartarchibald** [comment]
Noted out of band, electing to wait to merge https://github.com/IntelLabs/numba/pull/33 pending a more full investigation into `cell` support in inlining of closures with @ehsantn and @ninegua.

File: numba/tests/test_parfors.py [hunk 28]
```diff
@@ -497,6 +497,93 @@ def test_impl(x, is_positive, N):
...
+        def test_impl1(A):
+            return A.min()
+
+        def test_impl2(A):
+            return np.min(A)
+
+        n = 211
+        A = np.random.ranf(n)
+        B = np.random.randint(10, size=n).astype(np.int32)
+        C = np.random.ranf((n, n))  # test mutil-dimensional array
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/tests/test_parfors.py]: Done.

File: docs/source/user/parallel.rst [hunk 29]
```diff
@@ -67,8 +71,14 @@ Another experimental feature of this module is support for explicit parallel
 loops. One can use Numba's ``prange`` instead of ``range`` to specify that a
 loop can be parallelized. The user is required to make sure that the loop does
 not have cross iteration dependencies except the supported reductions.
-Currently, reductions on scalar values are supported and are inferred from
-in-place operations. The example below demonstrates a parallel loop with a
+
+A reductions is inferred automatically if a variable is updated by a binary
+function/operator using its previous value in the loop body. The initial value
+of the reduction is inferred automatically for ``+=`` and ``*=`` operators.
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply docs/source/user/parallel.rst]: Subtraction is not supported in reductions since it is not commutative. Supporting it could lead to substantial complexity. The user can easily use summation and subtract the final result.

File: docs/source/user/parallel.rst [hunk 30]
```diff
@@ -58,6 +59,9 @@ parallel semantics and for which we attempt to parallelize.
    Numpy broadcast between arrays with mixed dimensionality or size is
    not supported, nor is the reduction across a selected dimension.
 
+6. The ``reduce`` operator of ``functools`` is supported for specifying parallel
+   reductions on Numpy arrays but the initial value argument is mandatory.
+
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply docs/source/user/parallel.rst]: Added documentation for `reduce`. String constants are not mentioned in the documentation though.

File: docs/source/user/parallel.rst (hunk 29 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply docs/source/user/parallel.rst]: Ah yes, that's a fair point. Let's leave it.

File: numba/inline_closurecall.py [hunk 31]
```diff
@@ -54,10 +53,34 @@ def run(self):
                     lhs = instr.target
                     expr = instr.value
                     if isinstance(expr, ir.Expr) and expr.op == 'call':
+                        # inline reduce() when parallel is off
+                        if not self.flags.auto_parallel:
+                            call_name = guard(find_callname, self.func_ir, expr)
+                            if (call_name == ('reduce', 'builtin')
+                                    or call_name == ('reduce', 'functools')):
+                                assert len(expr.args) == 3, """invalid reduce
```
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/inline_closurecall.py]: I added checks for both `njit` and `parallel` modes.

**Comment by ehsantn** [comment]
Thanks @stuartarchibald . I addressed the new comments as well. Let me know if there are other issues.

File: numba/inline_closurecall.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/inline_closurecall.py]: Thanks very much. Could the sample above perhaps be used as a test?

**Comment by stuartarchibald** [comment]
@ehsantn thanks very much, I think the only other real issue is that the CI build is failing consistently e.g. https://travis-ci.org/numba/numba/jobs/275052317#L1138 as reduce is now supported and that test checks otherwise!

File: numba/inline_closurecall.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (ehsantn) [reply numba/inline_closurecall.py]: I just added the test.

**Comment by ehsantn** [comment]
I just fixed the test issue recently. The new builds should pass. Let's see :)

**Comment by stuartarchibald** [comment]
Thanks for all the fixes @ehsantn, much appreciated. I'm just pushing this through the much bigger build/test farm. Hopefully it'll pass and then we're good to go :)
